cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -203,15 +206,219 @@ pub enum SchemaType {
     },
 }
 
+impl<'de> Deserialize<'de> for SchemaType {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_any(SchemaTypeVisitor)
+    }
+}
+
+struct SchemaTypeVisitor;
+
+impl<'de> Visitor<'de> for SchemaTypeVisitor {
+    type Value = SchemaType;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("builtin type or reference to type defined in commonTypes")
+    }
+
+    fn visit_map<M>(self, mut map: M) -> std::result::Result<Self::Value, M::Error>
+    where
+        M: MapAccess<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(field_identifier, rename_all = "camelCase")]
+        enum Fields {
+            Type,
+            Element,
+            Attributes,
+            AdditionalAttributes,
+            Name,
+        }
+
+        let mut type_name: Option<SmolStr> = None;
+        let mut element: Option<SchemaType> = None;
+        let mut attributes: Option<BTreeMap<SmolStr, TypeOfAttribute>> = None;
+        let mut additional_attributes: Option<bool> = None;
+        let mut name: Option<SmolStr> = None;
+
+        while let Some(key) = map.next_key()? {
+            match key {
+                Fields::Type => {
+                    if type_name.is_some() {
+                        return Err(serde::de::Error::duplicate_field("type"));
+                    }
+                    type_name = Some(map.next_value()?)
+                }
+                Fields::Element => {
+                    if element.is_some() {
+                        return Err(serde::de::Error::duplicate_field("element"));
+                    }
+                    element = Some(map.next_value()?)
+                }
+                Fields::Attributes => {
+                    if attributes.is_some() {
+                        return Err(serde::de::Error::duplicate_field("attributes"));
+                    }
+                    attributes = Some(map.next_value()?)
+                }
+                Fields::AdditionalAttributes => {
+                    if additional_attributes.is_some() {
+                        return Err(serde::de::Error::duplicate_field("additionalAttributes"));
+                    }
+                    additional_attributes = Some(map.next_value()?)
+                }
+                Fields::Name => {
+                    if name.is_some() {
+                        return Err(serde::de::Error::duplicate_field("name"));
+                    }
+                    name = Some(map.next_value()?)
+                }
+            }
+        }
+
+        match type_name.as_ref().map(|s| s.as_str()) {
```

Comment:
[nit] I think this match statement could be `fn try_string_to_schema_type` or similar.